File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        //Bin search
        const int N = piles.size();
        
        //search space
        int low = 1, high = INT_MIN;
        for(int pile :piles) high = max(high, pile);
        
        int k = high;
        while(low <= high) {
            int mid = low +(high -low)/2;
            
            if(isPossibleEatBananaWithinHour(piles, h, mid)) {
                k = mid;
                high = mid -1;
            } else {
                low = mid +1;
            }
        }
        
        return k;
    }
    
private:
    bool isPossibleEatBananaWithinHour(vector<int>& piles, int h, int k) {
        int currHour = 0;
        for(int pile :piles) {
            currHour += ceil((double)pile /k);
            if(currHour > h) return false;
        }
        return true;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l = 0, r = 0;
        for(int p :piles) r = Math.max(r, p);
        
        int minTime = r;
        
        while(l <= r) {
            int k = (l +((r -l) /2));
            if(canEat(piles, h, k)) {
                minTime = k;
                r = k -1;
            } else {
                l = k +1;
            }
        }
        
        return minTime;
    }
    
    private boolean canEat(int[] piles, int h, int k) {
        int currTime = 0;
        
        for(int p :piles) {
            currTime += Math.ceil((double)p /k);
            if(currTime > h) return false;
        }
        return true;
    }
    
}